src/Specular/Internal/Incremental.purs

```diff
@@ -0,0 +1,410 @@
+          added <- runEffectFn2 PQ.add globalRecomputeQueue dependent
+--          if added then do
+--            dependentName <- runEffectFn1 Node.name dependent
+--            trace $ "stabilize: node " <> show dependentName <> " added to recompute queue"
+--          else do
+--            dependentName <- runEffectFn1 Node.name dependent
+--            trace $ "stabilize: node " <> show dependentName <> " already in recompute queue"
+          pure unit
+
+        observers <- runEffectFn1 Node.get_observers node
+        runEffectFn2 MutableArray.iterate observers $ mkEffectFn1 \observer -> do
+          -- FIXME: should be done outside stabilize loop, to avoid interfering with the process
+          -- (like in Specular - a FIFO queue)
+          runEffectFn1 observer newValue
+      else do
+--        trace $ "stabilize: node " <> show name <> " cut off"
+        pure unit
+
+  runEffectFn2 Ref.write globalCurrentStabilizationNum stabilizationIsNotInProgress
+--  trace "stabilize end"
+
+-- * Computational nodes
+
+constant :: forall a. EffectFn1 a (Node a)
+constant = mkEffectFn1 \value -> do
+  runEffectFn1 Node.create 
+    { compute: mkEffectFn1 \_ -> pure (Optional.some value)
+    , dependencies: pure []
+    }
+
+map :: forall a b. EffectFn2 (a -> b) (Node a) (Node b)
+map = mkEffectFn2 \fn a -> do
+  let deps = [toSomeNode a]
+  runEffectFn1 Node.create 
+    { compute: mkEffectFn1 \_ -> do
+        value_a <- runEffectFn1 Node.valueExc a
+        pure (Optional.some (fn value_a))
+    , dependencies: pure deps
+    }
+
+mapOptional :: forall a b. EffectFn2 (a -> Optional b) (Node a) (Node b)
+mapOptional = mkEffectFn2 \fn a -> do
+  let deps = [toSomeNode a]
+  runEffectFn1 Node.create 
+    { compute: mkEffectFn1 \_ -> do
+        value_a <- runEffectFn1 Node.get_value a
+        pure (if Optional.isSome value_a
+              then fn (Optional.fromSome value_a)
+              else Optional.none)
+    , dependencies: pure deps
+    }
+
+map2 :: forall a b c. EffectFn3 (Fn2 a b c) (Node a) (Node b) (Node c)
+map2 = mkEffectFn3 \fn a b -> do
+  let deps = [toSomeNode a, toSomeNode b]
+  runEffectFn1 Node.create 
+    { compute: mkEffectFn1 \_ -> do
+        value_a <- runEffectFn1 Node.valueExc a
+        value_b <- runEffectFn1 Node.valueExc b
+        pure (Optional.some (runFn2 fn value_a value_b))
+    , dependencies: pure deps
+    }
+
+-- Problem with bind and connect:
+-- We have to:
+-- - first compute LHS
+-- - then compute ourselves
+-- - only then we know the second dependency
+-- - compute the second dependency
+--
+-- Problem with bind and compute:
+-- - if we compute and the LHS changes, our height changes! We may need to re-add ourselves to recompute queue
+bind_ :: forall a b. EffectFn2 (Node a) (a -> Node b) (Node b)
+bind_ = mkEffectFn2 \lhs fn -> do
+  runEffectFn3 switch true lhs fn
+
+switch :: forall a b. EffectFn3 Boolean (Node a) (a -> Node b) (Node b)
+switch = mkEffectFn3 \alwaysFire lhs fn -> do
```

Comment:
Would be nice to have comments explaining which node is which. I understand this as such: 
`lsh <- rhs (created based on value resulting from applying f to value in lhs) <- main` (direction of dependencies)
Based on the value in "lhs" node  we create a new aux node "rhs" and update main - the output node - to depend on the new rhs node (removing its old dependencies and updating height to be at least height(rhs + 1))

Reply:
Agree, I also don't like the lhs/rhs terminology. Recently I prefer "control node"/"output node". Is that naming more clear?

Reply:
Yes.